barretenberg/cpp/src/barretenberg/polynomials/polynomial.hpp

```diff
@@ -21,8 +21,16 @@ template <typename Fr> struct PolynomialSpan {
     size_t end_index() const { return start_index + size(); }
     Fr* data() { return span.data(); }
     size_t size() const { return span.size(); }
-    Fr& operator[](size_t index) { return span[index - start_index]; }
-    const Fr& operator[](size_t index) const { return span[index - start_index]; }
+    Fr& operator[](size_t index)
+    {
+        ASSERT(index >= start_index && index < end_index());
```

Comment:
Adding some protection here as this can be quite unintuitive and got me a couple of times. E.g. If I have a "shiftable" poly like a wire or z_perm then I convert it to a PolynomialSpan, accessing the element at the 0th index silently returns undefined memory (as opposed to returning 0 or complaining as the original Polynomial would).